### PR TITLE
Install application CRD and add pipeline application CR to pipeline standalone

### PR DESCRIPTION
**Pull request by IronPan**
- Add application CRD and application manager
- Add pipeline application CR
- Set addOwnerRef=true for pipeline application CR so things will get GCed when application is deleted
- Set a common label `application-crd-id: kubeflow-pipelines` for identifying the KFP components. This will be used as identifier by application manager to set ownerref. Note we also set it for application manager themselves as they will get GCed too.

Note we don't set the label for CRD or namespace since they are cluster-scoped resources and application CRD doesn't support such resources. Setting labels on those resources will leads to unexpected GC behavior.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2585)

### CODE REVIEW

**Comment by IronPan** [comment]
fix https://github.com/kubeflow/pipelines/issues/2469

File: manifests/kustomize/base/kustomization.yaml [hunk 1]
```diff
@@ -2,24 +2,28 @@ apiVersion: kustomize.config.k8s.io/v1beta1
...
 - proxy
 - metadata
 
+resources:
+  - pipeline-application.yaml
+
 images:
 - name: gcr.io/ml-pipeline/api-server
-  newTag: 0.1.34
+  newTag: 0.1.33
```
Comment (Bobgy) [inline manifests/kustomize/base/kustomization.yaml]: Reminder, use new release: 0.1.34

File: manifests/kustomize/base/argo/minio-artifact-secret.yaml [hunk 2]
```diff
@@ -5,5 +5,6 @@ data:
 kind: Secret
 metadata:
   name: mlpipeline-minio-artifact
-  namespace: kubeflow
+  labels:
+    application-crd-id: kubeflow-pipelines
```
Comment (Bobgy) [inline manifests/kustomize/base/argo/minio-artifact-secret.yaml]: Looks like this needs to be added to every resource.

Can we add it by kustomization.yaml like `commonLabels` in https://github.com/kubernetes-sigs/kustomize/blob/master/examples/helloWorld/README.md#staging-kustomization?

File: manifests/kustomize/base/pipeline-application.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
+apiVersion: app.k8s.io/v1beta1
+kind: Application
```
Comment (Bobgy) [inline manifests/kustomize/base/pipeline-application.yaml]: Just confirmation:
* This is not GCP specific, is that correct?
* What value does this provide if not on GCP? Shall we consider putting it in an overlay?
* Should this be added to Kubeflow?

**Comment by Bobgy** [comment]
Also, note that there's an error in e2e tests when deploying these.
```
error: error validating "STDIN": error validating data: ValidationError(Deployment.spec): unknown field "serviceName" in io.k8s.api.apps.v1beta2.DeploymentSpec; if you choose to ignore these errors, turn validation off with --validate=false
```

File: manifests/kustomize/base/kustomization.yaml (hunk 1 above)
Comment (IronPan) [reply manifests/kustomize/base/kustomization.yaml]: done

File: manifests/kustomize/base/argo/minio-artifact-secret.yaml (hunk 2 above)
Comment (IronPan) [reply manifests/kustomize/base/argo/minio-artifact-secret.yaml]: thanks. done

File: manifests/kustomize/base/pipeline-application.yaml (hunk 3 above)
Comment (IronPan) [reply manifests/kustomize/base/pipeline-application.yaml]: this is not gcp specific but the goal of application CR is to provide a component to manage the monitor the application so i think this is still valuable for non gcp case. we should consider add to kubeflow manifest too.

File: manifests/kustomize/README.md [hunk 4]
```diff
@@ -6,8 +6,8 @@ This folder contains Kubeflow Pipelines Kustomize manifests for a light weight d
 
 Deploy latest version of Kubeflow Pipelines
 ```
-export PIPELINE_VERSION=0.1.31
-kubectl apply -f https://storage.googleapis.com/ml-pipeline/pipeline-lite/$PIPELINE_VERSION/namespaced-install.yaml
+export PIPELINE_VERSION=0.1.34
+for i in {1..2}; do kubectl apply -f https://storage.googleapis.com/ml-pipeline/pipeline-lite/$PIPELINE_VERSION/namespaced-install.yaml && break || sleep 1; done
```
Comment (Bobgy) [inline manifests/kustomize/README.md]: Can you help me understand why we need these retries in README?

Comment (Bobgy) [reply manifests/kustomize/README.md]: Besides, shall we sync this file with documentation on kubeflow/website?

File: manifests/kustomize/base/pipeline-application.yaml (hunk 3 above)
Comment (Bobgy) [reply manifests/kustomize/base/pipeline-application.yaml]: SG, I will take care of adding this to Kubeflow later.

File: test/deploy-pipeline-lite.sh [hunk 5]
```diff
@@ -48,7 +48,8 @@ kustomize edit set image gcr.io/ml-pipeline/visualization-server=${GCR_IMAGE_BAS
 kustomize edit set image gcr.io/ml-pipeline/inverse-proxy-agent=${GCR_IMAGE_BASE_DIR}/inverse-proxy-agent:${GCR_IMAGE_TAG}
 cat kustomization.yaml
 
-kustomize build . | kubectl apply -f -
+for i in {1..2}; do kustomize build . | kubectl apply -f - && break || sleep 1; done
```
Comment (Bobgy) [inline test/deploy-pipeline-lite.sh]: Why do we need to retry here? Where is flakiness?

File: manifests/kustomize/README.md (hunk 4 above)
Comment (IronPan) [reply manifests/kustomize/README.md]: adding crd definition is a eventual consistent behavior. by adding crd and cr in the same command, somehow master always throw me 
```
error: unable to recognize "STDIN": no matches for kind "Application" in version "app.k8s.io/v1beta1"
```
which is really pain. blind retry works so i added it here.

Comment (IronPan) [reply manifests/kustomize/README.md]: And regarding the doc, i can update once this is merged.

File: manifests/kustomize/base/pipeline-application.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    marketplace.cloud.google.com/deploy-info: '{"partner_id": "google-cloud-ai-platform", "product_id": "kubeflow-pipelines", "partner_name": "Google Cloud AI Platform"}'
+spec:
+  addOwnerRef: true
+  selector:
+    matchLabels:
+      application-crd-id: kubeflow-pipelines
+  descriptor:
+    version: 0.1.34
```
Comment (rmgogogo) [inline manifests/kustomize/base/pipeline-application.yaml]: Quick check, will our on-call script auto update this field? not familiar with that part.
It's OK if we put that change to another PR.

File: manifests/kustomize/base/crds/application-crd.yaml [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                  - type
+                  - status
+                type: object
+              type: array
+            observedGeneration:
+              format: int64
+              type: integer
+          type: object
+  version: v1beta1
+status:
```
Comment (Bobgy) [inline manifests/kustomize/base/crds/application-crd.yaml]: status should be data written from controller, I think we should remove them from here.
https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility

This fixes a minior issue: everytime you use `kubectl apply -k .`, it always shows
```
gcloud container clusters get-credentials test-2 --zone asia-east1-c --project gongyuan-pipeline-test configured
```
because the status part is overwritten

File: manifests/kustomize/base/pipeline-application.yaml (hunk 6 above)
Comment (Bobgy) [reply manifests/kustomize/base/pipeline-application.yaml]: We should update release playbook to also change this field

File: manifests/kustomize/README.md (hunk 4 above)
Comment (Bobgy) [reply manifests/kustomize/README.md]: > It might take a few seconds for the endpoint to be created. You can watch the Established condition of your CustomResourceDefinition to be true or watch the discovery information of the API server for your resource to show up.
https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition

I did some investigation, as you said, applying a CRD is eventual consistency (it takes some time before the k8s api endpoint is ready). So in reality it could take any amount of time before CRDs are ready. Using a retry here may work consistently now, but I think that's related to the time waited during retry just happen to be long enough for the CRD to be established. Any of these conditions may change in the future, e.g. if applying happens faster, or if CRDs take longer to establish, then this retry will break.

Also, everyone who read this the first time would wonder why we need a retry here.

I prefer we make the waiting explicit like the following. It is also pretty self explanatory about what we are doing.
```
kubectl apply -f crds
kubectl wait --for condition=established --timeout=60s crd/applications.app.k8s.io
kubectl apply -k envs/dev
```
https://stackoverflow.com/a/57126479

File: manifests/kustomize/base/crds/application-crd.yaml (hunk 7 above)
Comment (IronPan) [reply manifests/kustomize/base/crds/application-crd.yaml]: done

File: manifests/kustomize/base/pipeline-application.yaml (hunk 6 above)
Comment (IronPan) [reply manifests/kustomize/base/pipeline-application.yaml]: thanks. will do once this is merged.

**Comment by rmgogogo** [comment]
kindly ping this PR.
I also prefer change to apply twice, one for CRD and then for others to avoid retry logic.

File: manifests/kustomize/base/pipeline/ml-pipeline-viewer-crd-sa.yaml [hunk 8]
```diff
@@ -1,4 +1,4 @@
 apiVersion: v1
 kind: ServiceAccount
 metadata:
-  name: ml-pipeline-viewer-crd-service-account
+  name: ml-pipeline-viewer-crd-service-account
```
Comment (Bobgy) [inline manifests/kustomize/base/pipeline/ml-pipeline-viewer-crd-sa.yaml]: nit: can you revert this change?

File: test/manifests/crd/kustomization.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+# Actual image overrides will be added in test scripts.
+images: []
```
Comment (Bobgy) [inline test/manifests/crd/kustomization.yaml]: nit: these are not needed
